atomic_reactor/plugins/pre_pull_base_image.py

```diff
@@ -67,13 +67,17 @@ def run(self):
         Pull parent images and retag them uniquely for this build.
         """
         build_json = get_build_json()
-        base_image_str = str(self.workflow.builder.original_base_image)
         current_platform = platform.processor() or 'x86_64'
         self.manifest_list_cache = {}
         for nonce, parent in enumerate(sorted(self.workflow.builder.parent_images.keys())):
             image = ImageName.parse(parent)
-            if parent == base_image_str:
+            is_base_image = False
+            # original_base_image is an ImageName, so compare parent as an ImageName also
+            # since image is parent as an ImageName, just use that
+            if image == self.workflow.builder.original_base_image:
```

Comment:
No, I don't know how to trigger multistage builds.

Reply:
You can simply add another FROM instruction at the beginning of your Dockerfile. This can point to any other image we have available. There's also value in trying this out even without a multistage build.

Reply:
The following Dockerfile works:
```
FROM rhel7:7.2
FROM osbs-test/select-platforms
FROM  lucarval/rsyslog:7.2.lucarval

LABEL name=mlangsdo/hello-world \
      version=0.0.1 \
      com.redhat.component="chromium"
LABEL image=test
ENV x=y
RUN uname -a && env
```
relevant bits of the log for pull_base_image
2018-09-04 17:34:59,566 platform:- - atomic_reactor.plugin - DEBUG - running plugin 'pull_base_image'
2018-09-04 17:34:59,566 platform:- - atomic_reactor.plugin - INFO - running plugin instance with args: '{u'check_platforms': True}'
2018-09-04 17:34:59,566 platform:- - atomic_reactor.plugins.pull_base_image - INFO - using lucarval/rsyslog:7.2.lucarval as base image.
2018-09-04 17:34:59,566 platform:- - atomic_reactor.plugins.pull_base_image - INFO - Skipping validation of available platforms for base image because this is a single platform build
2018-09-04 17:34:59,567 platform:- - atomic_reactor.util - DEBUG - query_registry: querying /v2/lucarval/rsyslog/manifests/7.2.lucarval, headers: {u'Accept': 'application/vnd.docker.distribution.manifest.list.v2+json'}
2018-09-04 17:35:00,171 platform:- - atomic_reactor.util - DEBUG - query_registry: [302] http://brew-pulp-docker01.web.qa.ext.phx1.redhat.com:8888/v2/lucarval/rsyslog/manifests/7.2.lucarval
2018-09-04 17:35:00,171 platform:- - atomic_reactor.util - DEBUG - query_registry: [200] https://brew-pulp-docker01.web.qa.ext.phx1.redhat.com/pulp/docker/v2/redhat-lucarval-rsyslog/manifests/7.2.lucarval
2018-09-04 17:35:00,171 platform:- - atomic_reactor.util - DEBUG - query_registry: response headers: {'content-length': '5575', 'accept-ranges': 'bytes', 'keep-alive': 'timeout=5, max=100', 'server': 'Apache/2.4.6 (Red Hat Enterprise Linux) OpenSSL/1.0.2k-fips mod_wsgi/3.4 Python/2.7.5', 'last-modified': 'Tue, 10 Oct 2017 18:22:30 GMT', 'connection': 'Keep-Alive', 'etag': '"15c7-55b3563df5847"', 'docker-distribution-api-version': 'registry/2.0', 'date': 'Tue, 04 Sep 2018 17:35:00 GMT', 'content-type': 'application/x-troff-man'}
2018-09-04 17:35:00,171 platform:- - atomic_reactor.util - WARNING - content does not match expected media type
2018-09-04 17:35:00,172 platform:- - atomic_reactor.core - INFO - pulling image 'brew-pulp-docker01.web.qa.ext.phx1.redhat.com:8888/lucarval/rsyslog:7.2.lucarval' from registry
2018-09-04 17:35:00,172 platform:- - atomic_reactor.core - DEBUG - image = 'brew-pulp-docker01.web.qa.ext.phx1.redhat.com:8888/lucarval/rsyslog:7.2.lucarval', insecure = 'True'
2018-09-04 17:35:00,179 platform:- - atomic_reactor.util - INFO - wait_for_command
2018-09-04 17:35:11,575 platform:- - atomic_reactor.util - INFO - no more logs
2018-09-04 17:35:11,575 platform:- - atomic_reactor.plugins.pull_base_image - INFO - tagging pulled image
2018-09-04 17:35:11,576 platform:- - atomic_reactor.core - INFO - tagging image 'brew-pulp-docker01.web.qa.ext.phx1.redhat.com:8888/lucarval/rsyslog:7.2.lucarval' as 'scratch-29251-20180904173452:0'
2018-09-04 17:35:11,576 platform:- - atomic_reactor.core - DEBUG - image = 'brew-pulp-docker01.web.qa.ext.phx1.redhat.com:8888/lucarval/rsyslog:7.2.lucarval', target_image_name = 'scratch-29251-20180904173452:0'
2018-09-04 17:35:11,587 platform:- - atomic_reactor.plugins.pull_base_image - DEBUG - image 'brew-pulp-docker01.web.qa.ext.phx1.redhat.com:8888/lucarval/rsyslog:7.2.lucarval' is available as 'scratch-29251-20180904173452:0'
2018-09-04 17:35:11,587 platform:- - atomic_reactor.build - INFO - setting base image to 'scratch-29251-20180904173452:0'
2018-09-04 17:35:11,587 platform:- - atomic_reactor.plugins.pull_base_image - INFO - Skipping validation of available platforms for base image because this is a single platform build
2018-09-04 17:35:11,587 platform:- - atomic_reactor.util - DEBUG - query_registry: querying /v2/osbs-test/select-platforms/manifests/latest, headers: {u'Accept': 'application/vnd.docker.distribution.manifest.list.v2+json'}
2018-09-04 17:35:12,202 platform:- - atomic_reactor.util - DEBUG - query_registry: [302] http://brew-pulp-docker01.web.qa.ext.phx1.redhat.com:8888/v2/osbs-test/select-platforms/manifests/latest
2018-09-04 17:35:12,202 platform:- - atomic_reactor.util - DEBUG - query_registry: [200] https://brew-pulp-docker01.web.qa.ext.phx1.redhat.com/pulp/docker/v2/redhat-osbs-test-select-platforms/manifests/list/latest
2018-09-04 17:35:12,202 platform:- - atomic_reactor.util - DEBUG - query_registry: response headers: {'content-length': '470', 'accept-ranges': 'bytes', 'keep-alive': 'timeout=5, max=100', 'server': 'Apache/2.4.6 (Red Hat Enterprise Linux) OpenSSL/1.0.2k-fips mod_wsgi/3.4 Python/2.7.5', 'last-modified': 'Tue, 04 Sep 2018 03:46:57 GMT', 'connection': 'Keep-Alive', 'etag': '"1d6-57503827acfad"', 'docker-distribution-api-version': 'registry/2.0', 'date': 'Tue, 04 Sep 2018 17:35:12 GMT', 'content-type': 'application/vnd.docker.distribution.manifest.list.v2+json'}
2018-09-04 17:35:12,202 platform:- - atomic_reactor.util - DEBUG - content matches expected media type
2018-09-04 17:35:12,202 platform:- - atomic_reactor.core - INFO - pulling image 'brew-pulp-docker01.web.qa.ext.phx1.redhat.com:8888/osbs-test/select-platforms:latest' from registry
2018-09-04 17:35:12,202 platform:- - atomic_reactor.core - DEBUG - image = 'brew-pulp-docker01.web.qa.ext.phx1.redhat.com:8888/osbs-test/select-platforms:latest', insecure = 'True'
2018-09-04 17:35:12,210 platform:- - atomic_reactor.util - INFO - wait_for_command
2018-09-04 17:35:15,185 platform:- - atomic_reactor.util - INFO - no more logs
2018-09-04 17:35:15,185 platform:- - atomic_reactor.plugins.pull_base_image - INFO - tagging pulled image
2018-09-04 17:35:15,185 platform:- - atomic_reactor.core - INFO - tagging image 'brew-pulp-docker01.web.qa.ext.phx1.redhat.com:8888/osbs-test/select-platforms:latest' as 'scratch-29251-20180904173452:1'
2018-09-04 17:35:15,185 platform:- - atomic_reactor.core - DEBUG - image = 'brew-pulp-docker01.web.qa.ext.phx1.redhat.com:8888/osbs-test/select-platforms:latest', target_image_name = 'scratch-29251-20180904173452:1'
2018-09-04 17:35:15,204 platform:- - atomic_reactor.plugins.pull_base_image - DEBUG - image 'brew-pulp-docker01.web.qa.ext.phx1.redhat.com:8888/osbs-test/select-platforms:latest' is available as 'scratch-29251-20180904173452:1'
2018-09-04 17:35:15,204 platform:- - atomic_reactor.plugins.pull_base_image - INFO - Skipping validation of available platforms for base image because this is a single platform build
2018-09-04 17:35:15,205 platform:- - atomic_reactor.util - DEBUG - query_registry: querying /v2/rhel7/manifests/7.2, headers: {u'Accept': 'application/vnd.docker.distribution.manifest.list.v2+json'}
2018-09-04 17:35:15,720 platform:- - atomic_reactor.util - DEBUG - query_registry: [302] http://brew-pulp-docker01.web.qa.ext.phx1.redhat.com:8888/v2/rhel7/manifests/7.2
2018-09-04 17:35:15,721 platform:- - atomic_reactor.util - DEBUG - query_registry: [200] https://brew-pulp-docker01.web.qa.ext.phx1.redhat.com/pulp/docker/v2/redhat-rhel7/manifests/1/7.2
2018-09-04 17:35:15,721 platform:- - atomic_reactor.util - DEBUG - query_registry: response headers: {'content-length': '2178', 'accept-ranges': 'bytes', 'keep-alive': 'timeout=5, max=100', 'server': 'Apache/2.4.6 (Red Hat Enterprise Linux) OpenSSL/1.0.2k-fips mod_wsgi/3.4 Python/2.7.5', 'last-modified': 'Fri, 15 Jul 2016 19:52:42 GMT', 'connection': 'Keep-Alive', 'etag': '"882-537b1f6315019"', 'docker-distribution-api-version': 'registry/2.0', 'date': 'Tue, 04 Sep 2018 17:35:15 GMT', 'content-type': 'application/x-troff-man'}
2018-09-04 17:35:15,721 platform:- - atomic_reactor.util - WARNING - content does not match expected media type
2018-09-04 17:35:15,721 platform:- - atomic_reactor.core - INFO - pulling image 'brew-pulp-docker01.web.qa.ext.phx1.redhat.com:8888/rhel7:7.2' from registry
2018-09-04 17:35:15,721 platform:- - atomic_reactor.core - DEBUG - image = 'brew-pulp-docker01.web.qa.ext.phx1.redhat.com:8888/rhel7:7.2', insecure = 'True'
2018-09-04 17:35:15,728 platform:- - atomic_reactor.util - INFO - wait_for_command
2018-09-04 17:35:26,412 platform:- - atomic_reactor.util - INFO - no more logs
2018-09-04 17:35:26,412 platform:- - atomic_reactor.plugins.pull_base_image - INFO - tagging pulled image
2018-09-04 17:35:26,412 platform:- - atomic_reactor.core - INFO - tagging image 'brew-pulp-docker01.web.qa.ext.phx1.redhat.com:8888/rhel7:7.2' as 'scratch-29251-20180904173452:2'
2018-09-04 17:35:26,412 platform:- - atomic_reactor.core - DEBUG - image = 'brew-pulp-docker01.web.qa.ext.phx1.redhat.com:8888/rhel7:7.2', target_image_name = 'scratch-29251-20180904173452:2'
2018-09-04 17:35:26,422 platform:- - atomic_reactor.plugins.pull_base_image - DEBUG - image 'brew-pulp-docker01.web.qa.ext.phx1.redhat.com:8888/rhel7:7.2' is available as 'scratch-29251-20180904173452:2'

The untagged osbs-test/select-platforms is pulled as osbs-test/select-platforms:latest.

Reply:
Thank you for doing this testing.